### ISSUE TITLE
Private registries with no port are not properly handled

### ISSUE THREAD

**Issue by ldejager** [concourse/docker-image-resource#25]
Hello,

This issue can probably be seen as an extension to #17 in some sort of way.

The current project I am working on makes use of JFrog Bintray's hosted docker repositories and whilst it generally works, I'm also experiencing some problems.

Sample bintray repositories:

`companyrepos-docker-deploy-dev.bintray.io/containername`
`companyrepos-docker-deploy-dev.bintray.io/containername:latest`

Whilst I have approximately 10 Bintray resources defined (which works), I also rely on a public image being pulled from Docker Hub.

As per `cmd/check/main.go` the above listed sample repositories matches the second case in the code and as such the resource tries to look for the images on Docker Hub which hits the following code path:

`return officialRegistry, segs[0] + "/" + segs[1]`.

This along with the `check` resource type scripts means that Docker Hub gets hit multiple times for non Docker Hub hosted images which results into rate limiting and eventually `429 Too Many Requests` error messages when the pipeline tries to pull the Docker Hub hosted image. If I pause the pipeline for an hour, I'm able to resume docker push/pull operations against docker hub as normal.

The exact error message I get is:

```
resource script '/opt/resource/check []' failed: exit status 1

stderr:
failed to fetch manifest: Get https://registry-1.docker.io/v2/companyrepos-docker-deploy-dev.bintray.io/api/manifests/latest: token auth attempt for registry: https://auth.docker.io/token?account=username&scope=repository%3Acompanyrepos-docker-deploy-dev.bintray.io%2Fapi%3Apull&service=registry.docker.io request failed with status: 429 Too Many Requests
```

```
const officialRegistry = "registry-1.docker.io"

func parseRepository(repository string) (string, string) {
	segs := strings.Split(repository, "/")

	switch len(segs) {
	case 3:
		return segs[0], segs[1] + "/" + segs[2]
	case 2:
		if strings.Contains(segs[0], ":") {
			return segs[0], segs[1]
		} else {
			return officialRegistry, segs[0] + "/" + segs[1]
		}
	case 1:
		return officialRegistry, "library/" + segs[0]
	}

	fatal("malformed repository url")
	panic("unreachable")
}

```

Are there any plans to add better support for popular hosted repositories like bintray.io, quay.io?

Perhaps something like:

```
 name: git-resource-image
  type: docker-image
  source:
    provider: bintray.io
    repository: companyrepos-docker-deploy-dev.bintray.io/containername
    email: docker@example.com
    username: username
    password: password
```

Or anything else you feel appropriate. Locally I'm trying to work around the issue by adding a list of common docker repositories and then checking if what's provided in `repository` is in the list and if found, just returning that repo. It's not complete though as I'm trying to understand Go at the same time and would have much rather preferred to provide a patch than raise this issue.

There is a bintray-resource (https://github.com/jamiemonserrate/bintray-resource) though this only seems to be for non docker images.

**Comment by ldejager** [comment]
[fix_repository_flood.txt](https://github.com/concourse/docker-image-resource/files/173691/fix_repository_flood.txt)

The attached patch works somewhat - I still see a lot of requests to docker.io but nothing near the volumes I saw before which resulted into my account being temporarily blocked. I assume the requests I'm still seeing is perhaps from the `in` and `out` scripts.

There is probably a much more elegant and complete way to solve the issue so if this doesn't align with the project goals feel free to push back.

**Comment by cjcjameson** [comment]
@ldejager we had been trying to run against a private registry and ran into issues which seemed to manifest as an issue with specifying the port # in our docker url, but was actually an issue along this same codepath about understanding when a non-dockerhub registry is being used.

`my.private.docker.registry/image-name`: if we refer to the image this way, then pushing works, but all our "check" containers fail
`my.private.docker.registry:443/image-name`: if we refer to the image this way, the check containers work because they know not to refer to dockerhub. But any pushes we want to make fail.

We have a pipeline where we're chaining image creation from one job to the next, so we had no good way out. Fortunately, we just started using Dockerhub, so we're unblocked.

**Comment by dajulia3** [comment]
Would really love this change for our Pivotal Labs clients thinking of using concourse. For many of our enterprise customers we require private docker registries. For example, we have a docker image of a legacy db2 database seeded with realistic data, which they wouldn't want to be public. Right now we have a work-around, but it'd be nice to have this feature!

**Comment by DenisIzmaylov** [comment]
+1

**Comment by DenisIzmaylov** [comment]
I got the same issue, some of the logs:
```
Step 2 : COPY html /var/www/html
 ---> 627a205fcec0
Removing intermediate container f35cf1d025c5
Step 3 : EXPOSE 80
 ---> Running in d7d52ea162d9
 ---> 00c13e3f24d9
Removing intermediate container d7d52ea162d9
Step 4 : CMD /start.sh
 ---> Running in 8815e83217d0
 ---> 2ad0ca203f93
Removing intermediate container 8815e83217d0
Successfully built 2ad0ca203f93
The push refers to a repository [my-private-registry:5000/frontend-release]
Get https://my-private-registry:5000/v1/_ping: dial tcp 62.4.21.33:5000: i/o timeout
```

@dajulia3 Could you please share how did you make this work-around to get it possible to work with private Docker Registry?

**Comment by DenisIzmaylov** [comment]
Up!

**Comment by DenisIzmaylov** [comment]
I've launched Docker Registry without any complex config:
```
docker run -d -p 5000:5000 --restart=always --name registry \
  -v /var/docker-registry:/var/lib/registry \
  registry:2
```

**Comment by DenisIzmaylov** [comment]
It was `ufw` issue.

**Comment by vito** [comment]
Went with a pretty broad fix to just treat anything with a `.` in the first repository segment as a URL. Looking at Docker's code this is pretty much what they do, too. Fix should be in the next release.
